### ISSUE TITLE
All method doesn't work correctly with null values

### ISSUE THREAD

**Issue by samcook** [tumtumtum/Shaolinq#93]
If you have a DataAccessObject with a property that can contain nulls (e.g. string, nullable int, nullable bool, etc...), performing an `All(x => ...)` operation can give incorrect results, as it appears to generate SQL that ignores rows with null values in the field.

e.g.
```csharp
[DataAccessObject]
public class DbObject : DataAccessObject<int>
{
    [PersistedMember]
    public virtual string Foo { get; set; }
}
```
Id | Foo
-|-
1 | hello
2 | _NULL_
3 | _NULL_

```csharp
this.Model.DbObjects.All(x => x.Foo == "hello").ShouldBe(false);
// actually returns true

this.Model.DbObjects.Where(x => x.Foo == null).All(x => x.Foo == "anything").ShouldBe(false);
// actually returns true
```

The SQL generated for the first example looks something like:
```sql
SELECT
    (
        SELECT CASE
                   WHEN (NOT ((EXISTS
                               (
                                   SELECT "T0"."DbObjectId",
                                          "T0"."Foo"
                                   FROM "DbObject" AS "T0"
                                   WHERE NOT ((("T0"."Foo") = ('hello')))
                               )
                              )
                             )
                        ) THEN
            (1)
                   ELSE
            (0)
               END AS "EXISTS_COL"
    ) AS "value";
```

I think the problem is that the WHERE clause in the inner query is causing the rows with NULLs to be omitted.

**Comment by samcook** [comment]
Also affects `Any` with `!=`

```csharp
this.Model.DbObjects.Any(x => x.Foo != "hello").ShouldBe(false);
// actually returns true
```

**Comment by samcook** [comment]
For reference, this is what EntityFramework 6 generates:

```csharp
db.DbObjects.All(x => x.Foo == "hello"); // returns false
```
```sql
SELECT CASE
           WHEN (NOT EXISTS
                     (
                         SELECT 1 AS [C1]
                         FROM [dbo].[DbObjects] AS [Extent1]
                         WHERE (NOT (
                                        (N'hello' = [Extent1].[Foo])
                                        AND ([Extent1].[Foo] IS NOT NULL)
                                    )
                               )
                               OR (CASE
                                       WHEN (N'hello' = [Extent1].[Foo]) THEN
                                           CAST(1 AS BIT)
                                       WHEN (NOT (
                                                     (N'hello' = [Extent1].[Foo])
                                                     AND ([Extent1].[Foo] IS NOT NULL)
                                                 )
                                            ) THEN
                                           CAST(0 AS BIT)
                                   END IS NULL
                                  )
                     )
                ) THEN
               CAST(1 AS BIT)
           ELSE
               CAST(0 AS BIT)
       END AS [C1]
FROM
(SELECT 1 AS X) AS [SingleRowTable1];
```

```csharp
db.DbObjects.Any(x => x.Foo != "hello"); // returns true
```
```sql
SELECT CASE
           WHEN (EXISTS
                 (
                     SELECT 1 AS [C1]
                     FROM [dbo].[DbObjects] AS [Extent1]
                     WHERE NOT (
                                   (N'hello' = [Extent1].[Foo])
                                   AND ([Extent1].[Foo] IS NOT NULL)
                               )
                 )
                ) THEN
               CAST(1 AS BIT)
           ELSE
               CAST(0 AS BIT)
       END AS [C1]
FROM
(SELECT 1 AS X) AS [SingleRowTable1];
```

**Comment by tumtumtum** [comment]
Could you add a unit test?

**Comment by samcook** [comment]
Will do

**Comment by samcook** [comment]
Unsurprisingly, the same issue can also apply to the predicates for `Where`, `Count`, etc...

**Comment by tumtumtum** [comment]
TLDR but is this is an issue with SQL server only or all providers?

**Comment by samcook** [comment]
Affects both SQL server and sqlite, haven't tried mysql or postgres

**Comment by tumtumtum** [comment]
Yeah comparing with null is undefined so not(undefined) isn't true. 

I have to have a think about where else this affects could affect the framework because LINQ semantics for NULL is different.
